libdnf/advisory/advisory.cpp

```diff
@@ -0,0 +1,165 @@
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/advisory/advisory.hpp"
+
+#include "libdnf/advisory/advisory_collection.hpp"
+#include "libdnf/advisory/advisory_reference.hpp"
+#include "libdnf/common/exception.hpp"
+#include "libdnf/rpm/solv_sack_impl.hpp"
+#include "libdnf/utils/utils_internal.hpp"
+
+#include <fmt/format.h>
+#include <solv/pool.h>
+
+namespace libdnf::advisory {
+
+Advisory::Advisory(libdnf::rpm::SolvSack & sack, AdvisoryId id) : id(id), sack(sack.get_weak_ptr()) {}
+
+std::string Advisory::get_name() const {
+    const char * name;
+    Pool * pool = sack->p_impl->get_pool();
+
+    name = pool_lookup_str(pool, id.id, SOLVABLE_NAME);
+
+    if (strncmp(
+            libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX, name, libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH) !=
+        0) {
+        auto msg = fmt::format(
+            R"**(Bad libsolv id for advisory "{}", solvable name "{}" doesn't have advisory prefix "{}")**",
+            id.id,
+            name,
+            libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX);
+        throw RuntimeError(msg);
+    }
+
+    return std::string(name + libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH);
+}
+
+Advisory::Type Advisory::get_type() const {
+    const char * type;
+    Pool * pool = sack->p_impl->get_pool();
+    type = pool_lookup_str(pool, id.id, SOLVABLE_PATCHCATEGORY);
+
+    if (type == NULL) {
+        return Advisory::Type::UNKNOWN;
+    }
+    if (!strcmp(type, "bugfix")) {
+        return Advisory::Type::BUGFIX;
+    }
+    if (!strcmp(type, "enhancement")) {
+        return Advisory::Type::ENHANCEMENT;
+    }
+    if (!strcmp(type, "security")) {
+        return Advisory::Type::SECURITY;
+    }
+    if (!strcmp(type, "newpackage")) {
+        return Advisory::Type::NEWPACKAGE;
+    }
+
+    return Advisory::Type::UNKNOWN;
+}
+
+const char * Advisory::advisory_type_to_cstring(Type type) {
+    switch (type) {
+        case Type::ENHANCEMENT:
+            return "enhancement";
+        case Type::SECURITY:
+            return "security";
+        case Type::NEWPACKAGE:
+            return "newpackage";
+        case Type::BUGFIX:
+            return "bugfix";
+        default:
+            return NULL;
+    }
+}
+
+std::string Advisory::get_severity() const {
+    Pool * pool = sack->p_impl->get_pool();
+
+    //TODO(amatej): should we call SolvPrivate::internalize_libsolv_repo(solvable->repo);
+    //              before pool_lookup_str?
+    //              pool_lookup_str can return nullptr but std::string constructor cannot
+    //              accept nullptr (e.g. id = 0).
+    //              If so do this just once in solv::advisroy_private
+    return std::string(pool_lookup_str(pool, id.id, UPDATE_SEVERITY));
+}
+
+AdvisoryId Advisory::get_id() const {
+    return id;
```

Comment:
I suggest that AdvisoryId should have an explicit constructor with Id argument. Than conversion id (int) to AdvisoryId should not work.